### PR TITLE
21535 - Shortname Refund

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:*
https://github.com/bcgov/entity/issues/21535

*Description of changes:*
1. Data model changes - eft_refund
2. Data model changes - eft_refund_email_list (TODO when deploy to prod - add 4 expense authority emails)
3. Add schema for eft_refund
4. Add new route for eft_refund
5. Add email notification service

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

**Comment by Jxio** [comment]
I don't have updates on payment jobs
image

File: pay-api/src/pay_api/models/eft_refund_email_list.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    __tablename__ = 'eft_refund_email_list'
+    # this mapper is used so that new and old versions of the service can be run simultaneously,
+    # making rolling upgrades easier
+    # This is used by SQLAlchemy to explicitly define which fields we're interested
+    # so it doesn't freak out and say it can't map the structure if other fields are present.
+    # This could occur from a failed deploy or during an upgrade.
+    # The other option is to tell SQLAlchemy to ignore differences, but that is ambiguous
+    # and can interfere with Alembic upgrades.
+    #
+    # NOTE: please keep mapper names in alpha-order, easier to track that way
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund_email_list.py:35]: NOTE: please keep mapper names in alpha-order, easier to track that way

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 2]
```diff
@@ -189,3 +192,24 @@ def delete_eft_shortname_link(short_name_id: int, short_name_link_id: int):
...
+@bp.route('/shortname_refund', methods=['POST', 'OPTIONS'])
+@cross_origin(origins='*', methods=['POST'])
+@_jwt.has_one_of_roles(
+    [Role.SYSTEM.value, Role.EFT_REFUND.value])
+def post_shortname_refund():
+    """Create the Refund for the Shortname."""
+    current_app.logger.info('<post_shortname_refund')
+    request_json = request.get_json(silent=True)
+    try:
+        valid_format, errors = schema_utils.validate(request_json, 'refund_shortname') if \
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:206]: function? this is repeated alot

File: pay-api/src/pay_api/services/eft_service.py [hunk 3]
```diff
@@ -125,3 +133,45 @@ def create_receipt(invoice: InvoiceModel, payment: PaymentModel) -> ReceiptModel
                                              invoice_id=invoice.id,
                                              receipt_number=payment.receipt_number)
         return receipt
+
+    @classmethod
+    @user_context
+    def create_shortname_refund(cls, request: Dict[str, str], **kwargs) -> Dict[str, str]:
+        # pylint: disable=too-many-locals
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:140]: too many locals means you need to refactor the function

File: pay-api/src/pay_api/models/eft_refund_email_list.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from __future__ import annotations
+
+from sql_versioning import Versioned
+
+from .base_model import BaseModel
+from .db import db
+
+
+class EFTRefundEmailList(Versioned, BaseModel):
+    """This class manages all of the Expense Authority emails."""
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund_email_list.py:24]: Expense Authority emails that sounds fairly specific

File: pay-api/src/pay_api/models/eft_refund.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+from datetime import datetime
+
+from sqlalchemy import ForeignKey
+
+from .base_model import BaseModel
+from .db import db
+
+
+class EFTRefund(BaseModel):  # pylint: disable=too-many-instance-attributes
+    """This class manages the file data for EFT transactions."""
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund.py:24]: file data for EFT transactions? Does it? I thought this was just for EFT Refunds?

File: pay-api/src/pay_api/models/eft_refund_email_list.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    __mapper_args__ = {
+        'include_properties': [
+            'id',
+            'first_name',
+            'last_name',
+            'email'
+        ]
+    }
+
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund_email_list.py:46]: What are the relationships for this table? Seems fairly barebones now

File: pay-api/src/pay_api/models/eft_refund_email_list.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    first_name = db.Column(db.String(25), nullable=False)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/eft_refund_email_list.py:47]: What if they don't have a first name?

File: pay-api/tests/unit/models/test_eft_refund.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    refund_amount = 150.00
+    cas_supplier_number = 'SUP654321'
+    refund_email = 'updated@example.com'
+    comment = 'Updated comment'
+    status = 'COMPLETED'
+    updated_by = 'user123'
+    updated_by_name = 'User Name'
+    auth_account_id = '123'
+
+    eft_refund = EFTRefundModel()
```
Comment (seeker25) [inline pay-api/tests/unit/models/test_eft_refund.py:76]: Get rid of repetition of `eft_refund = `

why not
```
EFTRefundModel(
short_name_id=short_name_id,
refund_amount=refund_amount
....
)
```

File: pay-api/tests/unit/models/test_eft_refund.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+def test_eft_refund_defaults(session):
+    """Assert EFT refund defaults are stored."""
+    # Ensure the required entry exists in the related table
+    short_name = EFTShortnames(short_name='Test Short Name')
+    db.session.add(short_name)
+    db.session.commit()
+    short_name_id = short_name.id
+
+    # Create and save the EFT refund
```
Comment (seeker25) [inline pay-api/tests/unit/models/test_eft_refund.py:33]: redundant

File: pay-api/tests/unit/api/test_eft_short_names.py [hunk 10]
```diff
@@ -850,3 +851,42 @@ def test_apply_eft_short_name_credits(session, client, jwt, app):
...
+    data = {
+                'shortNameId': '12345',
+                'authAccountId': '123',
+                'refundAmount': 100.00,
+                'casSupplierNum': 'CAS123',
+                'refundEmail': 'test@example.com',
+                'comment': 'Refund for overpayment'
+            }
+
+    rv = client.post('/api/v1/eft-shortnames/shortname_refund', headers=headers, json=data)
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_short_names.py:873]: uhhh this doesn't look right

File: pay-api/tests/unit/api/test_eft_short_names.py [hunk 11]
```diff
@@ -850,3 +851,42 @@ def test_apply_eft_short_name_credits(session, client, jwt, app):
...
+def test_post_shortname_refund_invalid_request(client, mocker, jwt, app):
+    """Test handling of invalid request format."""
+    data = {
+        'invalid_field': 'invalid_value'
+    }
+
+    token = jwt.create_jwt(get_claims(roles=[Role.EFT_REFUND.value]), token_header)
+    headers = {'Authorization': f'Bearer {token}', 'content-type': 'application/json'}
+
+    rv = client.post('/api/v1/eft-shortnames/shortname_refund', headers=headers, json=data)
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_short_names.py:889]: this doesn't look right

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 12]
```diff
@@ -189,3 +192,24 @@ def delete_eft_shortname_link(short_name_id: int, short_name_link_id: int):
 
     current_app.logger.debug('>delete_eft_shortname_link')
     return jsonify(response), status
+
+
+@bp.route('/shortname_refund', methods=['POST', 'OPTIONS'])
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:197]: this doesn't look right

File: pay-api/src/pay_api/models/eft_refund_email_list.py (hunk 6 above)
Comment (Jxio) [reply pay-api/src/pay_api/models/eft_refund_email_list.py:46]: no it's standalone now, I only have list of names from the ticket

**Review by seeker25 — APPROVED** [review]
Looks great, well done

**Pull request merged**